Data/Euclidean.hs

```diff
@@ -271,3 +313,55 @@ instance Euclidean Double where
   quotRem x y = (x / y, 0)
   quot        = (/)
   rem         = const $ const 0
+
+instance Field Double
+
+instance GcdDomain CFloat where
+  divide x y = Just (x / y)
+  gcd        = const $ const 1
+  lcm        = const $ const 1
+  coprime    = const $ const True
+
+instance Euclidean CFloat where
+  degree      = const 0
+  quotRem x y = (x / y, 0)
+  quot        = (/)
+  rem         = const $ const 0
+
+instance Field CFloat
+
+instance GcdDomain CDouble where
+  divide x y = Just (x / y)
+  gcd        = const $ const 1
+  lcm        = const $ const 1
+  coprime    = const $ const True
+
+instance Euclidean CDouble where
+  degree      = const 0
+  quotRem x y = (x / y, 0)
+  quot        = (/)
+  rem         = const $ const 0
+
+instance Field CDouble
+
+instance (Eq a, Field a) => GcdDomain (Complex a) where
+  divide z (x :+ y)
+    | d == zero     = Nothing
```

Comment:
What justifies this special case? It is unexpected that `divide x 0 :+ 0 /= divide (x :+ 0) 0`. I'd expect that `divide` over a field always returns `Just smth`.

Reply:
Would it be better if this returned `Just (zero `quot` zero :+ zero `quot` zero)` as per `quot`?

Reply:
I think it is sufficient to specify only the general case here:
```haskell
divide z (x :+ y) = Just (z `times` ((x `quot` d) :+ (negate y `quot` d)))
```